### PR TITLE
[AI-BUILDER-FORM-SETUP-6]: add connection check to system prompt

### PR DESCRIPTION
**Pull request by m0nggh**
## Reinforce established FormSG connection context in system prompt

When a user connects a FormSG form early in a conversation, the LLM's recall of that connection (the `connection_id`, form title, and form ID) degrades over long conversations — even though the original message remains intact in every request. This causes the model to incorrectly emit connection pickers, claim the form URL is unconnected, or ask for a secret key for a trigger that is already fully verified.

To fix this, the system prompt is now augmented each turn with a server-derived reminder of the most recently established FormSG connection, extracted deterministically from the raw message history rather than relying on the model's own memory.

### What changed

- **`extractEstablishedFormConnection`** scans all user messages for the `I've connected my FormSG form "…" (id: …)` pattern (matching the exact text produced by the frontend's `buildFormConnectedMessage` / `buildKickoffMessage`) and returns the most recently mentioned connection's `connectionId`, `formId`, and `formTitle`.
- **`buildEstablishedConnectionReminder`** turns that extracted connection into a short, authoritative reminder appended to the system prompt, instructing the model to use the known `connection_id` directly and to skip any connection picker or secret-key prompts for this trigger.
- Both helpers are called on every request in the chat handler, keeping the fact visible near generation time regardless of conversation length.
- Unit tests cover the regex extraction (no match, full match with form ID, partial match without form ID, most-recent-wins, assistant messages ignored) and the reminder builder (null input, expected substrings).

<sub>Stack created with <a href="https://github.com/github/gh-stack">GitHub Stacks CLI</a> • <a href="https://gh.io/stacks-feedback">Give Feedback 💬</a></sub>

### CODE REVIEW

File: packages/backend/src/routes/api/chat/helpers.ts [hunk 1]
```diff
@@ -19,3 +19,69 @@ export function serializeMessagesForLangfuse(
...
+  connection: EstablishedFormConnection | null,
+): string {
+  if (!connection) {
+    return ''
+  }
+
+  const formRef = connection.formId ? ` (form id "${connection.formId}")` : ''
+
+  return `\n\nKnown fact, verified server-side just now — not from your own memory of earlier turns: the user already connected FormSG connection_id "${connection.connectionId}"${formRef} for the form "${connection.formTitle}" earlier in this conversation. When assigning this trigger's connection in Phase 2b step a, use this connection_id directly via update_step_parameters with parameter_labels: { "connection_id": "${connection.formTitle}" }. Do NOT emit a connection picker, do NOT say the form's URL is known but unconnected, and do NOT ask for a secret key for this trigger — this connection is already fully verified.`
+}
```
[the remark this answers is not in the document]
Comment (m0nggh) [reply packages/backend/src/routes/api/chat/helpers.ts:87]: !fp list_apps is a good safeguard to block the user from selecting the m365 excel app

**Comment by kevinkim-ogp** [comment]
*Posted by Claude*

Nit: `extractEstablishedFormConnection` scans forward and overwrites `latest` on every match, so it always walks the entire message array even after it's already found the match it'll end up using. `MAX_MESSAGES` (50) / `MAX_PARTS_PER_MESSAGE` (50) / `MAX_TEXT_LENGTH` (10k) already bound the worst case per request, so this isn't a real perf problem — but scanning backward and returning on the first match avoids the redundant work in the common case (the connect message is usually recent) with no behavior change:

```ts
export function extractEstablishedFormConnection(
  messages: ChatRequest['messages'],
): EstablishedFormConnection | null {
  for (let i = messages.length - 1; i >= 0; i--) {
    const message = messages[i]
    if (message.role !== 'user') {
      continue
    }
    for (let j = message.parts.length - 1; j >= 0; j--) {
      const part = message.parts[j]
      if (part.type !== 'text') {
        continue
      }
      const match = part.text.match(CONNECTED_FORM_MESSAGE_REGEX)
      if (match) {
        return {
          formTitle: match[1],
          connectionId: match[2],
          formId: match[3],
        }
      }
    }
  }

  return null
}
```

Non-blocking — up to you whether it's worth the diff.

**Comment by kevinkim-ogp** [comment]
*Posted by Claude*

## Code Review: [AI-BUILDER-FORM-SETUP-6]: add connection check to system prompt

**Mode:** Light (3 files, ~181 lines)
**Context:** Loaded root and backend `.claude/rules/*.md`; cross-checked the regex contract against `packages/frontend/src/pages/AiBuilder/helpers.ts` and the 24-hex FormSG form-id format used in `verify-credentials.ts`.

### Findings

#### 1. 🔵 Low · 🐛 Correctness
**Where:** `packages/backend/src/routes/api/chat/helpers.ts:56-63`
**Issue:** `extractEstablishedFormConnection` re-scans every user message's text on every request purely to find the last match, overwriting `latest` on each hit instead of stopping early.
**Fix:** See the separate comment above with a backward-scan version that returns on the first match — same semantics, less redundant work in the common case. Not blocking; `MAX_MESSAGES`/`MAX_PARTS_PER_MESSAGE`/`MAX_TEXT_LENGTH` already bound the worst case per request.

#### 2. ⚪ Info · 🛠️ Maintainability
**Where:** `packages/backend/src/routes/api/chat/helpers.ts:86`
**Issue:** The reminder hardcodes `"Phase 2b step a"` and the exact `update_step_parameters`/`parameter_labels` call shape, coupling this backend string to prompt structure that presumably lives in the Langfuse-hosted prompt, not in this repo. If the prompt's phase numbering changes, this reminder silently goes stale with no compiler/test signal.
**Fix:** Consider a comment pointing to wherever the Langfuse prompt's "Phase 2b" is authored/versioned, so both are updated together.

### Summary

**Verdict:** Approve

**Top priorities:** none blocking.

**Done well:** The regex is well-anchored against the actual producer (`buildFormConnectedMessage`/`buildKickoffMessage` in the frontend), and the test suite in `helpers.test.ts` covers the real edge cases that matter here — no match, form-id-present vs absent, most-recent-wins, and assistant messages being excluded (correctly preventing the model's own echoed text from being treated as a verified fact).

**Review by kevinkim-ogp — APPROVED** [review]
lgtm!
don't have to address claude's comments now. can monitor and park for enhancement

File: packages/backend/src/routes/api/chat/helpers.ts [hunk 2]
```diff
@@ -19,3 +19,69 @@ export function serializeMessagesForLangfuse(
...
+          formTitle: match[1],
+          connectionId: match[2],
+          formId: match[3],
+        }
+      }
+    }
+  }
+
+  return null
+}
```
[the remark this answers is not in the document]
Comment (m0nggh) [reply packages/backend/src/routes/api/chat/helpers.ts:68]: ok will address it in another pr
